exchanges/gateio/ws_ob_update_manager.go

```diff
@@ -0,0 +1,152 @@
+		cache.buffer = append(cache.buffer, updateWithFirstID{update: update, firstID: firstID})
+		return nil
+	}
+
+	if book == nil || book.LastUpdateID+1 < firstID /*orderbook is behind notifications so refresh*/ {
+		cache.updating = true
+		go func() {
+			// REST Orderbook IDs are behind the ws update feed by about 50-100 changes; inline rudimentary delay to cache
+			time.Sleep(time.Second)
+			if err := cache.updateOrderbookAndApply(ctx, g, update.Pair, update.Asset, limit); err != nil {
+				g.Websocket.DataHandler <- err
+			}
+		}()
+		cache.buffer = append(cache.buffer, updateWithFirstID{update: update, firstID: firstID})
+		return nil
+	}
+
+	if book.LastUpdateID+1 > update.UpdateID {
+		return fmt.Errorf("%w for LastUpdateID:%d while baseID+1:%d", errOutOfOrder, update.UpdateID, book.LastUpdateID+1)
+	}
+
+	return cache.applyUpdate(g, update)
+}
+
+func (m *wsOBUpdateManager) getCache(p currency.Pair, a asset.Item) *updateCache {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	cache, ok := m.m[key.PairAsset{Base: p.Base.Item, Quote: p.Quote.Item, Asset: a}]
+	if !ok {
+		cache = &updateCache{}
+		m.m[key.PairAsset{Base: p.Base.Item, Quote: p.Quote.Item, Asset: a}] = cache
+	}
+	return cache
+}
+
+func (c *updateCache) updateOrderbookAndApply(ctx context.Context, g *Gateio, pair currency.Pair, a asset.Item, limit uint64) error {
+	defer c.mtx.Unlock()
+	defer func() { c.buffer = nil; /*gc buffer; unused until snapshot required*/ c.updating = false }()
+	book, err := g.UpdateOrderbookWithLimit(ctx, pair, a, limit)
+	c.mtx.Lock() // lock here to prevent ws handle data interference with REST request above
+	if err != nil {
+		return err
+	}
+
+	if a != asset.Spot {
+		err = g.Websocket.Orderbook.LoadSnapshot(book)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Spot, Margin, and Cross Margin books are the same
+		for _, a := range standardMarginAssetTypes {
+			if enabled, _ := g.CurrencyPairs.IsPairEnabled(pair, a); !enabled {
+				continue
+			}
+			book.Asset = a
+			err = g.Websocket.Orderbook.LoadSnapshot(book)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return c.applyPending(g, a)
+}
+
+func (c *updateCache) applyPending(g *Gateio, a asset.Item) error {
+	for _, data := range c.buffer {
+		book, err := g.Websocket.Orderbook.GetOrderbook(data.update.Pair, a)
+		if err != nil {
+			return err
+		}
+		nextID := book.LastUpdateID + 1
+		if data.firstID > nextID {
+			return errOrderbookSnapshotBehind
+		}
+
+		if data.update.UpdateID < nextID {
+			continue // skip updates that are behind the current orderbook
+		}
+		if err := c.applyUpdate(g, data.update); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *updateCache) applyUpdate(g *Gateio, update *orderbook.Update) error {
+	if update.Asset != asset.Spot {
+		return g.Websocket.Orderbook.Update(update)
+	}
+
+	for _, a := range standardMarginAssetTypes {
+		if enabled, _ := g.CurrencyPairs.IsPairEnabled(update.Pair, a); !enabled {
+			continue
```

Comment:
Would checking the map be better? Or is it chicken and egg? These seem like more expensive checks when you know exactly which pairs and assets are being done

Reply:
Not a bad idea; only issue I see is to the persistence of the map I think it would be funky between assets if we say drop one margin pair, it will still think its enabled.

Reply:
That's all good then